File: leetcode/301-400/301-remove-invalid-parentheses/bench/parenrepair.c
/* Benchmark mirror of parenrepair.kara — LeetCode #301, unique-by-construction
 * repair. Same recursion, same depth-indexed scratch buffer, same sink.
 *
 * `scratch[32][32]` is what the Kara mirror spells as one flat Vec[u8] indexed
 * by depth * 32; keeping the shape identical is what makes the comparison
 * algorithm-for-algorithm rather than allocator-for-allocator. */
#include <stdio.h>

#define NCASES 2000
#define SLEN   24
#define PASSES 64
#define SLOT   32
#define MAXDEPTH 32
#define MOD 1000000007LL

static unsigned char corpus[NCASES * SLEN];
static unsigned char scratch[MAXDEPTH][SLOT];
static long long results;
static long long checksum;

static void repair(int depth, int len, int last_i, int last_j,
                   unsigned char open, unsigned char close) {
    unsigned char *base = scratch[depth];
    unsigned char *child = scratch[depth + 1];

    int count = 0;
    for (int i = last_i; i < len; i++) {
        unsigned char c = base[i];
        if (c == open) count++;
        else if (c == close) count--;
        if (count < 0) {
            for (int j = last_j; j <= i; j++) {
                if (base[j] == close && (j == last_j || base[j - 1] != close)) {
                    int w = 0;
                    for (int k = 0; k < len; k++)
                        if (k != j) child[w++] = base[k];
                    repair(depth + 1, len - 1, i, j, open, close);
                }
            }
            return;
        }
    }

    for (int r = 0; r < len; r++) child[r] = base[len - 1 - r];

    if (open == '(') {
        repair(depth + 1, len, 0, 0, ')', '(');
    } else {
        long long h = 0;
        for (int t = 0; t < len; t++) h = (h * 31 + child[t]) % MOD;
        results++;
        checksum = (checksum + h) % MOD;
    }
}

int main(void) {
    long long state = 12345;
    for (int n = 0; n < NCASES * SLEN; n++) {
        state = (state * 1103515245LL + 12345LL) & 0x7fffffffLL;
        long long r = (state / 65536) % 3;
        corpus[n] = r == 0 ? '(' : (r == 1 ? ')' : 'a');
    }

    for (int p = 0; p < PASSES; p++) {
        for (int ci = 0; ci < NCASES; ci++) {
            const unsigned char *src = corpus + (size_t)ci * SLEN;
            for (int k = 0; k < SLEN; k++) scratch[0][k] = src[k];
            repair(0, SLEN, 0, 0, '(', ')');
        }
    }

    printf("results %lld\n", results);
    printf("checksum %lld\n", checksum);
    return 0;
}

File: leetcode/301-400/301-remove-invalid-parentheses/bench/parenrepair.rs
// Benchmark mirror of parenrepair.kara — LeetCode #301, unique-by-construction
// repair. Same recursion, same depth-indexed scratch buffer, same sink.
//
// The scratch buffer is one flat Vec<u8> indexed by `depth * SLOT`, exactly as
// the Kara mirror spells it — not a `[[u8; 32]; 32]` split into disjoint
// borrows. Two reasons: it keeps the five mirrors byte-for-byte the same
// algorithm, and it keeps Rust's bounds checks on the same accesses Kara
// checks, which is what makes the equal-safety column honest.

const NCASES: usize = 2000;
const SLEN: usize = 24;
const PASSES: usize = 64;
const SLOT: usize = 32;
const MAXDEPTH: usize = 32;
const MOD: i64 = 1_000_000_007;

#[allow(clippy::too_many_arguments)]
fn repair(
    scratch: &mut [u8],
    depth: usize,
    len: usize,
    last_i: usize,
    last_j: usize,
    open: u8,
    close: u8,
    results: &mut i64,
    checksum: &mut i64,
) {
    let base = depth * SLOT;
    let child = base + SLOT;

    let mut count: i64 = 0;
    let mut i = last_i;
    while i < len {
        let c = scratch[base + i];
        if c == open {
            count += 1;
        } else if c == close {
            count -= 1;
        }
        if count < 0 {
            let mut j = last_j;
            while j <= i {
                if scratch[base + j] == close
                    && (j == last_j || scratch[base + j - 1] != close)
                {
                    let mut w = 0;
                    for k in 0..len {
                        if k != j {
                            scratch[child + w] = scratch[base + k];
                            w += 1;
                        }
                    }
                    repair(scratch, depth + 1, len - 1, i, j, open, close, results, checksum);
                }
                j += 1;
            }
            return;
        }
        i += 1;
    }

    for r in 0..len {
        scratch[child + r] = scratch[base + len - 1 - r];
    }

    if open == b'(' {
        repair(scratch, depth + 1, len, 0, 0, b')', b'(', results, checksum);
    } else {
        let mut h: i64 = 0;
        for t in 0..len {
            h = (h * 31 + scratch[child + t] as i64) % MOD;
        }
        *results += 1;
        *checksum = (*checksum + h) % MOD;
    }
}

fn main() {
    let mut corpus = vec![0u8; NCASES * SLEN];
    let mut state: i64 = 12345;
    for slot in corpus.iter_mut() {
        state = (state * 1103515245 + 12345) & 0x7fff_ffff;
        let r = (state / 65536) % 3;
        *slot = if r == 0 { b'(' } else if r == 1 { b')' } else { b'a' };
    }

    let mut scratch = vec![0u8; MAXDEPTH * SLOT];
    let mut results: i64 = 0;
    let mut checksum: i64 = 0;

    for _ in 0..PASSES {
        for ci in 0..NCASES {
            let src = ci * SLEN;
            scratch[..SLEN].copy_from_slice(&corpus[src..src + SLEN]);
            repair(&mut scratch, 0, SLEN, 0, 0, b'(', b')', &mut results, &mut checksum);
        }
    }

    println!("results {results}");
    println!("checksum {checksum}");
}
